src/main/java/com/redhat/devtools/lsp4ij/LSPIJEditorUtils.java

```diff
@@ -0,0 +1,419 @@
+    }
+
+    /**
+     * Returns the brace pair close character for the specified brace pair open character in the provided file.
+     *
+     * @param file               the PSI file
+     * @param openBraceCharacter the optional brace pair open character
+     * @return the corresponding brace pair close character if found; otherwise null
+     */
+    @Nullable
+    @ApiStatus.Internal
+    public static Character getCloseBraceCharacter(@NotNull PsiFile file, @Nullable Character openBraceCharacter) {
+        return openBraceCharacter != null ? getBracePairsFwd(file).get(openBraceCharacter) : null;
+    }
+
+    @NotNull
+    private static Map<Character, Character> getBracePairsFwd(@NotNull PsiFile file) {
+        return CachedValuesManager.getCachedValue(file, () -> CachedValueProvider.Result.create(getBracePairs(file), file));
+    }
+
+    @NotNull
+    private static Map<Character, Character> getBracePairsBwd(@NotNull PsiFile file) {
+        return CachedValuesManager.getCachedValue(file, () -> CachedValueProvider.Result.create(ContainerUtil.reverseMap(getBracePairsFwd(file)), file));
+    }
+
+    // Comments
+
+    @NotNull
+    public static Commenter getCommenter(@NotNull PsiFile file) {
+        Commenter commenter = null;
+
+        if (isAbstractFileTypeFile(file)) {
+            AbstractFileType abstractFileType = (AbstractFileType) file.getFileType();
+            commenter = abstractFileType.getCommenter();
+        }
+
+        // NOTE: Unfortunately the language grammar comment information is not available for TextMate bundles
+
+        if (commenter == null) {
+            commenter = new LSPIJDefaultCommenter(file);
+        }
+
+        return commenter;
+    }
+
+    private static class LSPIJDefaultCommenter implements Commenter {
+        private final String lineCommentPrefix;
+        private final String blockCommentPrefix;
+        private final String blockCommentSuffix;
+
+        private LSPIJDefaultCommenter(@NotNull PsiFile file) {
+            LSPClientFeatures clientFeatures = getClientFeatures(file);
+            this.lineCommentPrefix = clientFeatures != null ? clientFeatures.getLineCommentPrefix(file) : null;
+            this.blockCommentPrefix = clientFeatures != null ? clientFeatures.getBlockCommentPrefix(file) : null;
+            this.blockCommentSuffix = clientFeatures != null ? clientFeatures.getBlockCommentSuffix(file) : null;
+        }
+
+        @Override
+        public @Nullable String getLineCommentPrefix() {
+            return lineCommentPrefix;
+        }
+
+        @Override
+        public @Nullable String getBlockCommentPrefix() {
+            return blockCommentPrefix;
+        }
+
+        @Override
+        public @Nullable String getBlockCommentSuffix() {
+            return blockCommentSuffix;
+        }
+
+        @Override
+        public @Nullable String getCommentedBlockCommentPrefix() {
+            return null;
```

Comment:
We don't care about these two for purposes of identifying whether or not we're in a block comment.